solidity/random-beacon/contracts/libraries/Groups.sol

```diff
@@ -343,40 +344,46 @@ library Groups {
         return shiftedIndex;
     }
 
-    /// @notice Sets addresses of members for the group eliminating members at
-    ///         positions pointed by the misbehavedMembersIndices array.
-    ///
-    ///         NOTE THAT THIS FUNCTION CHANGES ORDER OF MEMBERS IN THE GROUP
-    ///         IF THERE IS AT LEAST ONE MISBEHAVED MEMBER
-    ///
-    ///         The final group members indexes should be obtained post-DKG
-    ///         and they may differ from the ones outputted by the group
-    ///         selection protocol.
+    /// @notice Sets a hash of group members that participated in a group signing
+    ///         key generation. This function filters out IA/DQ members before
+    ///         hashing.
     /// @param group The group storage.
     /// @param members Group member addresses as outputted by the group selection
     ///        protocol.
     /// @param misbehavedMembersIndices Array of misbehaved (disqualified or
     ///        inactive) group members. Indices reflect positions
     ///        of members in the group as outputted by the group selection
-    ///        protocol.
-    function setGroupMembers(
+    ///        protocol. Indices must be in ascending order.
+    function setGroupMembersHash(
         Group storage group,
         uint32[] calldata members,
         uint8[] calldata misbehavedMembersIndices
     ) private {
-        group.members = members;
-
-        // Iterate misbehaved array backwards, replace misbehaved
-        // member with the last element and reduce array length
-        uint256 i = misbehavedMembersIndices.length;
-        while (i > 0) {
+        if (misbehavedMembersIndices.length > 0) {
+            // members that generated a group signing key
+            uint32[] memory groupMembers = new uint32[](
+                members.length - misbehavedMembersIndices.length
+            );
+            uint256 k = 0; // misbehaved members counter
+            uint256 j = 0; // group members counter
             // group member indices start from 1, so we need to -1 on misbehaved
-            uint8 memberArrayPosition = misbehavedMembersIndices[i - 1] - 1;
-            group.members[memberArrayPosition] = group.members[
-                group.members.length - 1
-            ];
-            group.members.pop();
-            i--;
+            uint8 misbehavedMemberArrayPosition = misbehavedMembersIndices[k] -
+                1;
+            for (uint256 i = 0; i < members.length; i++) {
+                if (i != misbehavedMemberArrayPosition) {
+                    groupMembers[j] = members[i];
+                    j++;
+                } else if (k < misbehavedMembersIndices.length - 1) {
+                    k++;
+                    misbehavedMemberArrayPosition =
+                        misbehavedMembersIndices[k] -
+                        1;
+                }
+            }
```

Comment:
Would this work fine if we get rid of `misbehavedMemberArrayPosition` like in suggestion below?
```suggestion
            for (uint256 i = 0; i < members.length; i++) {
                if (i != misbehavedMembersIndices[k] - 1) {
                    groupMembers[j] = members[i];
                    j++;
                } else if (k < misbehavedMembersIndices.length - 1) {
                    k++;
                }
            }
```

Reply:
Yup 08c8a4b